File: datasets/Part 2 - Regression/polinomial_regression.py
# -*- coding: utf-8 -*-
"""
Created on Thu May  5 11:00:48 2022

@author: mudar
"""

# Regresion Polinomica

# Como importar las librerias
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importar el dataset
dataset = pd.read_csv('data/Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values # para que no sea un vector y sea una matriz ponemos 1:2
y = dataset.iloc[:, 2].values


# En este caso no vamos a dividir el conjunto de entrenamiento en test train y
# y test dada la escasa cantidad de datos

# Dividir el dataset en conjunto de entrenamiento y conjunto de testing
# 80% conjunto entrenamiento y 20% conjunto de testing
"""
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
"""


# Escalado de variables
# Para que no haya tanta diferencia en los rangos de
# valores en los datos de las distintas columnas
"""
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
"""


# Ajustar la Regresion Lineal con todo el dataset para comparar
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, y)

# Ajustar la Regresion Polinomica con todo el dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y)


# Visualizacion de los resultados del Modelo Lineal
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_reg.predict(X), color = 'blue')
plt.title('Modelo de Regresion Lineal')
plt.xlabel('Posicion del empleado')
plt.ylabel('Sueldo (en $)')
plt.show()

# Visualizacion de los resultados del Modelo Polinomico
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, lin_reg_2.predict(poly_reg.fit_transform(X_grid)), color = 'blue')
plt.title('Modelo de Regresion Polinomica')
plt.xlabel('Posicion del empleado')
plt.ylabel('Sueldo (en $)')
plt.show()


# Prediccion de nuestros modelos
# En el siguiente caso con el modelo lineal sale un valor irreal
lin_reg.predict([[6.5]])

# Con el modelo polinomico si que se ajusta mas a los datos
lin_reg_2.predict(poly_reg.fit_transform([[6.5]]))

File: datasets/Part 2 - Regression/polinomial_regression.R

# Regresion Polinomica

setwd("~/repos/machinelearning-az/datasets/Part 2 - Regression")

# Importar el dataset
dataset = read.csv('data/Position_Salaries.csv')
dataset = dataset[, 2:3]

# En este caso no se va a dividir en conjunto de entrenamiento y test
# debido a la escasa cantidad de datos

# Dividir los datos en conjunto de entrenamiento y conjunto de test
# library(caTools) # Instalar con install.packages('caTools')
# set.seed(123) # Para que salgan los mismos resultados que en el curso
# split = sample.split(dataset$Purchased, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# testing_set = subset(dataset, split == FALSE)

# Escalado de valores
#training_set[, 2:3] = scale(training_set[, 2:3])
#testing_set[, 2:3] = scale(testing_set[, 2:3])


# Ajustar Modelo de Regresion Lineal con el Conjunto de Datos
lin_reg = lm(formula = Salary ~ .,
             data = dataset)

# Ajustar Modelo de Regresion Polinomica con el Conjunto de Datos
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ .,
              data = dataset)

# Visualizacion del Modelo Lineal
library(ggplot2) # Instalar con install.packages('ggplot2')
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            color = 'blue') +
  ggtitle('Prediccion lineal del sueldo en funcion del nivel del empleado') +
  xlab('Nivel del empleado') +
  ylab('Sueldo (en $)')

# Visualizacion del Modelo Polinomico
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = 'red') +
  geom_line(aes(x = x_grid, y = predict(poly_reg, newdata = data.frame(Level = x_grid,
                                                                       Level2 = x_grid^2,
                                                                       Level3 = x_grid^3,
                                                                       Level4 = x_grid^4))),
            color = 'blue') +
  ggtitle('Prediccion polinomica del sueldo en funcion del nivel del empleado') +
  xlab('Nivel del empleado') +
  ylab('Sueldo (en $)')

# Subiendo el grado del polinomio la curva de prediccion se adapta mejor a los puntos

# Prediccion de nuevos resultados con Regresion Lineal
y_pred = predict(lin_reg, newdata = data.frame(Level = 6.5))

# Prediccion de nuevos resultados con Regresion Polinomica
y_pred_poly = predict(poly_reg, newdata = data.frame(Level = 6.5,
                                                     Level2 = 6.5^2,
                                                     Level3 = 6.5^3,
                                                     Level4 = 6.5^4))
